bigquery: classes to support external table config

jba opened the pull request:
These classes will be used to configure external table
definitions for queries (transient external tables)
as well as external data configurations for tables (permanent
external tables).

This PR just establishes the classes. Subsequent PRs will
wire them into query jobs and tables.

tswast commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,485 @@
+        if schema:
+            config.schema = _parse_schema_resource(schema)
+        for optcls in (BigtableOptions, CSVOptions, GoogleSheetsOptions):
+            opts = resource.pop(optcls._RESOURCE_NAME, None)
```
Clever!

tswast commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,485 @@
+    """
+
+    qualifier_encoded = _TypedApiResourceProperty(
+        'qualifier_encoded', 'qualifierEncoded', six.binary_type)
```
Does `six.binary_type` mean that we are handling the conversion to base-64 for them?

We should clarify in the docstring whether we expect binary or base-64.

jba replied:
Done.

tswast commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,485 @@
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#externalDataConfiguration.bigtableOptions.columnFamilies.columns.qualifier_encoded
+    """
+
+    qualifier_string = _TypedApiResourceProperty(
```
We could potentially do like you do for options and take care of populating the right field depending on if we are given a binary value or a unicode value.

jba replied:
Yes, but it's very subtle because of the different binary/text types in Python 2 vs. 3. For example, if someone passes in `str`, that will be binary in Python 2 but text in Python 3.

Probably safer to have less magic.

tswast replied:
We do a little bit of this kind of magic in the DB-API,

https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/bigquery/google/cloud/bigquery/dbapi/_helpers.py#L49

But, I'm okay erring on the side of less magic.

tswast commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,485 @@
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#externalDataConfiguration.bigtableOptions.columnFamilies.columns.qualifier_string
+    """
+
+    type = _TypedApiResourceProperty('type', 'type', six.string_types)
```
Since [`type` is a Python built-in](https://docs.python.org/2/library/functions.html#type), the convention is a trailing underscore, like `type_`.

tswast commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,485 @@
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#externalDataConfiguration.bigtableOptions.columnFamilies.onlyReadLatest
+    """
+
+    type = _TypedApiResourceProperty('type', 'type', six.string_types)
```
Ditto on using `type_` here.

jba replied:
Done.

jba commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,485 @@
+        self._schema = value
+
+    @property
+    def options(self):
```
BQ is still using an HTTP/JSON discovery doc. I don't think there is anything like oneof.

jba commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,485 @@
+
+    @schema.setter
+    def schema(self, value):
+        if not all(isinstance(field, SchemaField) for field in value):
```
I was copying from LoadJobConfig, which has an additional constraint. Changed.

jba commented:
In addition to addressing your comments, I made source_format a required arg of `ExternalConfig` (seemed natural to me as I started to use it), and implemented checking of source format vs. options.

jba commented:
I just had another idea: if you write
```
ec = ExternalConfig('BIGTABLE')
```
we could populate options with an empty `BigtableOptions`, so you could just start setting
```
ec.options.something_for_bigtable = foo
```
That way we're not setting the user up for failure by having to pick the right XXXOptions class.

WDYT?

tswast commented:
Your suggestion on populating options in the constructor makes sense to me.

tswast commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,491 @@
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#externalDataConfiguration.maxBadRecords
+    """
+
+    source_format = _TypedApiResourceProperty(
```
Do we want to make this read-only now that we're setting it in the constructor?

jba replied:
Done.

tswast commented on bigquery/google/cloud/bigquery/external_config.py:
```diff
@@ -0,0 +1,491 @@
+    """
+
+    @property
+    def options(self):
```
If you do set this in the constructor, we could make this read-only, too, but keep the options mutable.

jba replied:
Will do in next PR.
